### PR TITLE
Feature: cache first fetching

### PR DESCRIPTION
**Pull request by Kurczaak**
What's been done:
- Added a boolean parameter `isCacheFirst` to `fetchWithin` and `fetchWithinDistance`
- `_collectionFutures` based on the `isCacheFirst` flag tries to fetch data from cache and if no result matches the query then the server is the source of truth

### CODE REVIEW

**Comment by kosukesaigusa** [comment]
@Kurczaak 

Thank you for creating a PR! 

At first glance it seems good, but let met check later (maybe tomorrow)!

File: lib/src/geo_collection_reference.dart [hunk 1]
```diff
@@ -282,15 +288,28 @@ class GeoCollectionReference<T> {
...
+        try {
+          querySnapshot = await query.get(
+            GetOptions(
+              source: isCacheFirst ? Source.cache : Source.serverAndCache,
+            ),
+          );
+        } on FirebaseException catch (_) {
+          if (isCacheFirst) {
+            querySnapshot = await query.get();
+          }
```
Comment (kosukesaigusa) [inline lib/src/geo_collection_reference.dart:311]: When `isCacheFirst` is `false`, the `FirebaseException` caught here is ignored, so how about `rethrow` it in `else` phrase?

```suggestion
          if (isCacheFirst) {
            querySnapshot = await query.get();
          } else {
            rethrow;
          }
```

**Review by kosukesaigusa — COMMENTED** [review]
@Kurczaak 

Thank you for your PR!

I added a comment, please take a look!

File: lib/src/geo_collection_reference.dart (hunk 1 above)
Comment (Kurczaak) [reply lib/src/geo_collection_reference.dart]: Hi, great remark, I haven't noticed that, thanks!
Can you have a look at the refactored version from my latest commit? I used your suggestion and refactored `catch` section a bit so it's more concise now.

Comment (kosukesaigusa) [reply lib/src/geo_collection_reference.dart]: @Kurczaak 

Cool! CI unit test also passes, so I will approve now!

**Review by kosukesaigusa — APPROVED** [review]
@Kurczaak 

Thank you so much for your contribution!

I will merge this PR, and release new version with this change soon!

**Comment by kosukesaigusa** [comment]
@Kurczaak 

The new geoflutterfire_plus version v0.0.23, includes your contribution has been released!

https://pub.dev/packages/geoflutterfire_plus/changelog

000098@2x

**Pull request merged**